### PR TITLE
Move state logic to state.rs

### PR DESCRIPTION
**Pull request by andagaev**

### CODE REVIEW

File: crates/oct-cloud/src/state.rs [hunk 1]
```diff
@@ -13,17 +13,144 @@ pub struct Ec2InstanceState {
...
+            public_dns: Some(self.public_dns.clone()),
+            region: self.region.clone(),
+            ami: self.ami.clone(),
+            instance_type,
+            name: self.name.clone(),
+            user_data: "".to_string(),
+            user_data_base64: "".to_string(),
+            instance_profile,
+        })
+    }
```
Comment (minev-dev) [inline crates/oct-cloud/src/state.rs:71]: You don't need this logic, it's already implemented in `aws::Ec2Instance::new`.

The same idea should be followed in the `InstanceProfileState` and `InstanceRoleState`

File: crates/oct-cloud/src/aws.rs [hunk 2]
```diff
@@ -637,18 +577,32 @@ mod tests {
                     .instances(
                         aws_sdk_ec2::types::Instance::builder()
                             .instance_id("id")
-                            .public_ip_address("1.1.1.1")
-                            .public_dns_name("example.com")
-                            .outpost_arn("arn")
+                            .public_ip_address(String::new())
+                            .public_dns_name(String::new())
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:581]: Should they be `None`?

File: crates/oct-cloud/src/aws.rs [hunk 3]
```diff
@@ -523,15 +490,15 @@ impl Resource for InstanceProfile {
 
 #[derive(Debug)]
 pub struct InstanceRole {
-    client: IAM,
+    pub client: IAM,
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:493]: `client` shouldn't be public, it's an implementation detail

File: crates/oct-cloud/src/aws.rs [hunk 4]
```diff
@@ -422,13 +427,13 @@ impl Resource for Ec2Instance {
 
 #[derive(Debug)]
 pub struct InstanceProfile {
-    client: IAM,
+    pub client: IAM,
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:430]: `client` shouldn't be public, it's an implementation detail

File: crates/oct-cloud/src/aws.rs [hunk 5]
```diff
@@ -396,9 +356,55 @@ impl Resource for Ec2Instance {
...
+        }
+
+        if self.public_ip.is_none() || self.public_dns.is_none() {
+            return Err("Failed to retrieve instance metadata after retries".into());
+        }
+
+        log::info!(
+            "Metadata retrieved: public_ip={}, public_dns={}",
+            self.public_ip.clone().unwrap(),
+            self.public_dns.clone().unwrap(),
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:406]: Don't use `unwrap` in the production code. It'll cause panic

File: crates/oct-cloud/src/aws.rs [hunk 6]
```diff
@@ -396,9 +356,55 @@ impl Resource for Ec2Instance {
...
+                        }
+                    }
+                }
+            } else {
+                return Err("Instance ID is None during metadata polling".into());
+            }
+
+            attempts += 1;
+            tokio::time::sleep(SLEEP_DURATION).await;
+        }
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:397]: What do you think about moving this logic to a private method?

File: crates/oct-cloud/src/aws.rs [hunk 7]
```diff
@@ -396,9 +356,55 @@ impl Resource for Ec2Instance {
...
+                                self.public_ip = Some(public_ip.to_string());
+                            }
+                            if let Some(public_dns) = instance.public_dns_name() {
+                                self.public_dns = Some(public_dns.to_string());
+                            }
+
+                            // Break if all metadata is available
+                            if self.public_ip.is_some() && self.public_dns.is_some() {
+                                break;
+                            }
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:387]: Is it possible to have only one of them not None? I suppose if one of them is available, another one should also be available

File: crates/oct-cloud/src/aws.rs [hunk 8]
```diff
@@ -396,9 +356,55 @@ impl Resource for Ec2Instance {
             .ok_or("No instances returned")?;
 
         self.id = instance.instance_id.clone();
-        self.arn = instance.outpost_arn.clone();
-        self.public_ip = instance.public_ip_address.clone();
-        self.public_dns = instance.public_dns_name.clone();
+
+        log::info!("Instance launched, waiting for metadata to be available");
+
+        tokio::time::sleep(std::time::Duration::from_secs(30)).await;
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:362]: Do we really need this sleep if we already have enough number of attempts getting the metadata?

File: crates/oct-cloud/src/aws.rs [hunk 9]
```diff
@@ -396,9 +356,55 @@ impl Resource for Ec2Instance {
...
+        const MAX_ATTEMPTS: usize = 10;
+        const SLEEP_DURATION: std::time::Duration = std::time::Duration::from_secs(5);
+
+        while attempts < MAX_ATTEMPTS {
+            if let Some(instance_id) = &self.id {
+                if let Ok(describe_response) =
+                    self.client.describe_instances(instance_id.clone()).await
+                {
+                    if let Some(reservation) = describe_response.reservations().get(0) {
+                        if let Some(instance) = reservation.instances().get(0) {
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:375]: Can you avoid using so deep nesting?

File: crates/oct-cloud/src/aws.rs [hunk 10]
```diff
@@ -231,34 +245,32 @@ impl IAMImpl {
...
-use IAMImpl as IAM;
+pub use IAMImpl as IAM;
 #[cfg(test)]
-use MockIAMImpl as IAM;
+pub use MockIAMImpl as IAM;
 
 #[derive(Debug)]
 pub struct Ec2Instance {
-    client: Ec2,
+    pub client: Ec2,
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:254]: `client` shouldn't be public, it's an implementation detail

File: crates/oct-cloud/src/aws.rs [hunk 11]
```diff
@@ -231,34 +245,32 @@ impl IAMImpl {
 }
 
 #[cfg(not(test))]
-use IAMImpl as IAM;
+pub use IAMImpl as IAM;
 #[cfg(test)]
-use MockIAMImpl as IAM;
+pub use MockIAMImpl as IAM;
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:250]: These clients shouldn't be public, it's an implementation detail

File: crates/oct-cloud/src/aws.rs [hunk 12]
```diff
@@ -83,19 +97,19 @@ impl Ec2Impl {
...
-struct IAMImpl {
+pub struct IAMImpl {
     inner: aws_sdk_iam::Client,
 }
 
 /// TODO: Add tests using static replay
 #[cfg_attr(test, automock)]
 impl IAMImpl {
-    fn new(inner: aws_sdk_iam::Client) -> Self {
+    pub fn new(inner: aws_sdk_iam::Client) -> Self {
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:112]: It shouldn't be public, it's for internal use only

File: crates/oct-cloud/src/aws.rs [hunk 13]
```diff
@@ -83,19 +97,19 @@ impl Ec2Impl {
...
+pub use Ec2Impl as Ec2;
 #[cfg(test)]
-use MockEc2Impl as Ec2;
+pub use MockEc2Impl as Ec2;
 
 #[derive(Debug)]
-struct IAMImpl {
+pub struct IAMImpl {
     inner: aws_sdk_iam::Client,
 }
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:107]: This code shouldn't be public, it's for internal use only

File: crates/oct-cloud/src/aws.rs [hunk 14]
```diff
@@ -25,16 +26,29 @@ pub trait Resource {
...
-struct Ec2Impl {
+pub struct Ec2Impl {
     inner: aws_sdk_ec2::Client,
 }
 
 /// TODO: Add tests using static replay
 #[cfg_attr(test, automock)]
 impl Ec2Impl {
-    fn new(inner: aws_sdk_ec2::Client) -> Self {
+    pub fn new(inner: aws_sdk_ec2::Client) -> Self {
```
Comment (minev-dev) [inline crates/oct-cloud/src/aws.rs:36]: This code shouldn't be public, it's for internal use only

**Review by minev-dev — CHANGES_REQUESTED** [review]

**Review by minev-dev — APPROVED** [review]

**Review by olegasics — APPROVED** [review]

**Pull request merged**